### PR TITLE
Avoid unnecessary rebuild on local installation

### LINKED ISSUE #2143: `cargo install --path .` recompiles
`cargo install --path .` recompiles the current crate in `target-install` before installing. This makes cargo install useless for package managers that want to separate build from package logic, wastes time, and wastes space (even small rust projects put a lot of data into the target directory). Cargo should just re-use the current target directory.

### PR DESCRIPTION
**Pull request by JIghtuse**
Fixes #2143

### CODE REVIEW

File: tests/test_cargo_install.rs [hunk 1]
```diff
@@ -543,3 +543,24 @@ test!(installs_from_cwd_by_default {
...
+            [package]
+            name = "foo"
+            version = "0.1.0"
+            authors = []
+        "#)
+        .file("src/main.rs", "fn main() {}");
+    assert_that(p.cargo_process("build").arg("--release"),
+                execs().with_status(0));
+
+    assert_that(cargo_process("install").arg("--path").arg(p.root()),
```
Comment (alexcrichton) [inline tests/test_cargo_install.rs]: Ah here you can use `p.cargo("install")` to avoid blowing away the directory as `p.cargo_process` does

**Comment by alexcrichton** [comment]
Looks great to me, thanks @JIghtuse!

File: tests/test_cargo_install.rs (hunk 1 above)
Comment (JIghtuse) [inline tests/test_cargo_install.rs]: Hm, am I misunderstand something? Build directory stays alive after installation after my patch. Why should we switch to `p.cargo("install")`. Anyway, I added more tests to make sure build directory and release binary is alive.

Comment (alexcrichton) [inline tests/test_cargo_install.rs]: Ah yeah sorry I just meant that `p.cargo(...)` was more idiomatic, but for `install` it's probably fine to just use this (as it's what the other install tests do)

**Pull request merged**